File: example/cpp/tools/research/whole_body_rolling_horizon_probe.cpp
// Offline fixed12ms scheduler: frozen12ms prefix + optimized20ms tail.
// Simulation advances after synchronous computation; measured lateness only gates
// adoption. This is not a realtime worker and grants no production authority.
#include "../../terrain/stage_c/whole_body_force_tracking.h"
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sys/file.h>
#include <fcntl.h>
#include <unistd.h>
using namespace stage_c_research;
using Q=Eigen::Matrix<double,19,1>;using V=Eigen::Matrix<double,18,1>;
using Gain=Eigen::Matrix<double,12,36,Eigen::RowMajor>;
struct Ref{Q q;V v;Torque u;Gain K;};
struct Stage{int tick;Ref law;FootForces forces;double solve_ms;std::string status;};
struct Plan{int version=0,source_version=-1,observed_tick=0,start_tick=0,end_tick=6;double generation_ms=0;bool valid=false,adopted=false,late=false;std::string failure;std::vector<Stage> stages;std::vector<double> initial_state;};
struct Row{int tick,version;Torque tau;FootForces forces;std::vector<double> next_state;};
template<class T> void Read(std::istream& in,T& x){for(int i=0;i<x.size();++i)in>>x.data()[i];if(!in || !x.allFinite())throw std::runtime_error("invalid input array");}
template<class T> void Print(const T& x){std::cout<<'[';for(int i=0;i<x.size();++i){if(i)std::cout<<',';std::cout<<x.data()[i];}std::cout<<']';}
std::vector<double> State(const mjModel* m,const mjData* d){std::vector<double> x(mj_stateSize(m,mjSTATE_INTEGRATION));mj_getState(m,d,x.data(),mjSTATE_INTEGRATION);return x;}
Torque Control(const mjModel* m,const mjData* d,const Ref& ref){Eigen::Matrix<double,36,1> e;mj_differentiatePos(m,e.data(),1,ref.q.data(),d->qpos);for(int j=0;j<18;++j)e[18+j]=d->qvel[j]-ref.v[j];return ref.u-ref.K*e;}
Ref AtState(const mjData* d,const Torque& u,const Gain& K){Ref r;r.q=Eigen::Map<const Q>(d->qpos);r.v=Eigen::Map<const V>(d->qvel);r.u=u;r.K=K;return r;}
void Advance(const mjModel* m,mjData* d,const Torque& u){for(int j=0;j<12;++j)d->ctrl[j]=u[j];mj_step(m,d);}
int main(int argc,char** argv){
 if(argc!=2)return 2;int lock=open("/tmp/go2_mujoco_experiment.lock",O_CREAT|O_WRONLY,0600);if(lock<0 || flock(lock,LOCK_EX|LOCK_NB))return 3;
 try{
  std::ifstream in(argv[1]);std::string magic,scene;std::getline(in,magic);std::getline(in,scene);if(magic!="whole-body-rolling-horizon-input-v1")throw std::runtime_error("invalid magic");
  char error[2048];std::unique_ptr<mjModel,decltype(&mj_deleteModel)> model(mj_loadXML(scene.c_str(),nullptr,error,sizeof(error)),mj_deleteModel);if(!model)throw std::runtime_error(error);auto m=model.get();
  int count=0,N=0;double vx=0,period=0;in>>count>>N>>vx>>period;if(count!=mj_stateSize(m,mjSTATE_INTEGRATION) || N!=70 || m->nq!=19 || m->nv!=18 || m->nu!=12 || std::abs(m->opt.timestep-.002)>1e-15 || std::abs(period-.14)>1e-12)throw std::runtime_error("requires Go2 2ms 70step reference");
  std::vector<double> initial(count);for(double& v:initial)in>>v;for(double v:initial)if(!std::isfinite(v))throw std::runtime_error("nonfinite initial state");
  std::vector<Ref> refs(N);for(auto& r:refs){Read(in,r.q);Read(in,r.v);Read(in,r.u);Read(in,r.K);}std::string end;in>>end;if(end!="end" || in>>end)throw std::runtime_error("invalid input end");
  std::unique_ptr<mjData,decltype(&mj_deleteData)> actual(mj_makeData(m),mj_deleteData),predicted(mj_makeData(m),mj_deleteData);auto d=actual.get();auto p=predicted.get();
  mj_setState(m,d,initial.data(),mjSTATE_INTEGRATION);d->qvel[1]+=.05;initial=State(m,d);double start_time=d->time;WholeBodyForceTracker tracker(m,4);
  auto nominal=[&](int tick){Ref r=refs[tick%N];r.q[0]+=(tick/N)*vx*period;return r;};
  Plan bootstrap;bootstrap.valid=true;bootstrap.adopted=true;bootstrap.start_tick=0;bootstrap.end_tick=6;
  for(int k=0;k<6;++k)bootstrap.stages.push_back({k,nominal(k),FootForces::Zero(),0,"bootstrap_nominal_feedback"});
  std::vector<Plan> plans;plans.reserve(240);std::vector<Row> rows;rows.reserve(1400);Plan active=bootstrap;int pending=-1;std::string failure;int failure_tick=-1;
  auto law_at=[&](const Plan& plan,int tick)->const Ref&{if(tick<plan.start_tick || tick>=plan.end_tick)throw std::runtime_error("law outside coverage");for(const Stage& s:plan.stages)if(s.tick==tick)return s.law;throw std::runtime_error("law sample missing");};
  for(int tick=0;tick<1400;++tick){
   if(tick%6==0){
    // Adoption is ordered before the next observation. No publication can occur
    // while that observation's six-step commitment is being executed.
    if(pending>=0){Plan& candidate=plans[pending];if(candidate.valid && !candidate.late && candidate.source_version==active.version){candidate.adopted=true;active=candidate;}pending=-1;}
    if(tick+6<1400){
     if(active.end_tick<tick+6){failure="missing_committed_prefix";failure_tick=tick;break;}
     Plan candidate;candidate.version=static_cast<int>(plans.size())+1;candidate.source_version=active.version;candidate.observed_tick=tick;candidate.start_tick=tick+6;candidate.end_tick=tick+16;candidate.initial_state=State(m,d);
     auto begin=std::chrono::steady_clock::now();mj_setState(m,p,candidate.initial_state.data(),mjSTATE_INTEGRATION);
     for(int j=0;j<16;++j){
      int future=tick+j;Ref ref=j<6?law_at(active,future):nominal(future);Torque requested=Control(m,p,ref);ForceTrackingResult solved;
      if(j<6){solved.torque=requested.cwiseMax(-35).cwiseMin(35);tracker.SetState(p);solved.forces=tracker.Evaluate(solved.torque);solved.feasible=solved.forces.maxCoeff()<=180;solved.status="frozen_published_prefix";}
      else solved=tracker.Solve(p,requested);
      candidate.stages.push_back({future,AtState(p,solved.torque,ref.K),solved.forces,solved.elapsed_ms,solved.status});
      if(!solved.feasible){candidate.failure=(j<6?"committed_prefix_force_violation:":"optimized_tail_failure:")+solved.status;break;}
      Advance(m,p,solved.torque);
     }
     candidate.valid=candidate.stages.size()==16 && candidate.failure.empty();candidate.generation_ms=std::chrono::duration<double,std::milli>(std::chrono::steady_clock::now()-begin).count();candidate.late=candidate.generation_ms>12;
     plans.push_back(std::move(candidate));pending=static_cast<int>(plans.size())-1;
     if(!plans.back().valid){failure=plans.back().failure;failure_tick=tick;break;}
    }
   }
   if(tick>=active.end_tick){failure="active_law_expired";failure_tick=tick;break;}
   Torque tau=Control(m,d,law_at(active,tick)).cwiseMax(-35).cwiseMin(35);tracker.SetState(d);FootForces f=tracker.Evaluate(tau);Advance(m,d,tau);rows.push_back({tick,active.version,tau,f,State(m,d)});
   if(f.maxCoeff()>180 || tau.cwiseAbs().maxCoeff()>35){failure="actual_force_or_torque_violation";failure_tick=tick;break;}
  }
  std::cout<<std::setprecision(17)<<"{\"schema\":\"whole-body-continuous-rolling-horizon-v1\",\"scope\":\"offline initialized flat; explicitly authorized periodic nominal reuse; synchronous simulator with measured late discard, not liveRT\",\"start_time\":"<<start_time<<",\"target_steps\":1400,\"completed_steps\":"<<rows.size()<<",\"failure_tick\":"<<failure_tick<<",\"failure\":\""<<failure<<"\",\"update_period_steps\":6,\"committed_prefix_steps\":6,\"optimized_tail_steps\":10,\"initial_integration_state\":";Print(initial);
  std::cout<<",\"candidates\":[";
  for(size_t index=0;index<plans.size();++index){const Plan& c=plans[index];if(index)std::cout<<',';std::cout<<"{\"version\":"<<c.version<<",\"source_version\":"<<c.source_version<<",\"observation_tick\":"<<c.observed_tick<<",\"start_tick\":"<<c.start_tick<<",\"end_tick\":"<<c.end_tick<<",\"generation_ms\":"<<c.generation_ms<<",\"valid\":"<<(c.valid?"true":"false")<<",\"late\":"<<(c.late?"true":"false")<<",\"adopted\":"<<(c.adopted?"true":"false")<<",\"failure\":\""<<c.failure<<"\",\"initial_integration_state\":";Print(c.initial_state);std::cout<<",\"stages\":[";
   for(size_t k=0;k<c.stages.size();++k){const Stage& s=c.stages[k];if(k)std::cout<<',';std::cout<<"{\"tick\":"<<s.tick<<",\"qpos\":";Print(s.law.q);std::cout<<",\"qvel\":";Print(s.law.v);std::cout<<",\"tau\":";Print(s.law.u);std::cout<<",\"K\":";Print(s.law.K);std::cout<<",\"forces\":";Print(s.forces);std::cout<<",\"solve_ms\":"<<s.solve_ms<<",\"status\":\""<<s.status<<"\"}";}
   std::cout<<"]}";
  }
  std::cout<<"],\"rows\":[";for(size_t k=0;k<rows.size();++k){const Row& r=rows[k];if(k)std::cout<<',';std::cout<<"{\"tick\":"<<r.tick<<",\"active_version\":"<<r.version<<",\"tau\":";Print(r.tau);std::cout<<",\"forces\":";Print(r.forces);std::cout<<",\"next_integration_state\":";Print(r.next_state);std::cout<<'}';}std::cout<<"]}\n";
 }catch(const std::exception& e){std::cerr<<e.what()<<'\n';return 1;}
 close(lock);return 0;
}

File: example/cpp/tools/research/whole_body_rolling_horizon_probe.py
"""Hash-bound input and independent continuous replay for one fixed12ms schedule.
Periodic nominal reuse is authorized for this initialized offline diagnostic only.
"""
import argparse,copy,fcntl,pathlib
import mujoco,numpy as np
import whole_body_horizon_probe as probe

def prepare(packet_dir,out):
    manifest,refs,nominal=probe.packet(packet_dir);out=pathlib.Path(out)
    if len(refs)!=70: raise ValueError('requires the 70step periodic research reference')
    fmt=lambda a:' '.join(format(float(x),'.17g') for x in np.asarray(a).ravel())
    lines=['whole-body-rolling-horizon-input-v1',manifest['model']['scene'],f"{len(nominal['initial_integration_state'])} 70 {nominal['command_vx']:.17g} {nominal['period_s']:.17g}",fmt(nominal['initial_integration_state'])]
    for r in refs: lines.append(fmt(np.r_[r['qpos'],r['qvel'],r['tau'],r['K'].ravel()]))
    with (out/'input.txt').open('x') as f:f.write('\n'.join(lines+['end'])+'\n')
    probe.write(out/'input_manifest.json',{'packet_dir':str(pathlib.Path(packet_dir).resolve()),'manifest_sha256':probe.sha(pathlib.Path(packet_dir)/'manifest.json'),'packet_sha256':probe.sha(pathlib.Path(packet_dir)/'trajectory.packet'),'input_sha256':probe.sha(out/'input.txt'),'offline_periodic_reference_reuse_authorized':True,'update_steps':6,'prefix_steps':6,'tail_steps':10,'periods':20,'initial_vy_delta':.05,'late_policy':'discard; stop if remaining old coverage cannot supply full 12ms commitment'})

def can_commit(end_tick,observation_tick): return end_tick>=observation_tick+6
def audit_schedule(run):
    """Validate epoch/coverage/late records independently of physics replay."""
    candidates=run['candidates'];by_tick={};by_version={0:{'start_tick':0,'end_tick':6}}
    for index,c in enumerate(candidates):
        obs=c['observation_tick']
        if c['version']!=index+1 or obs%6 or obs in by_tick:raise ValueError('candidate identity/calendar')
        if c['start_tick']!=obs+6 or c['end_tick']!=obs+16:raise ValueError('candidate coverage')
        if not np.isfinite(c['generation_ms']) or c['generation_ms']<0 or c['late']!=(c['generation_ms']>12):raise ValueError('invalid measured lateness')
        if c['valid'] and (len(c['stages'])!=16 or [s['tick'] for s in c['stages']]!=list(range(obs,obs+16))):raise ValueError('candidate sample coverage')
        by_tick[obs]=c;by_version[c['version']]=c
    rows=run['rows'];active=0;adopted=set();last=max(len(rows)-1,max(by_tick,default=-1))
    for tick in range(last+1):
        if tick%6==0:
            prior=by_tick.get(tick-6)
            if prior and prior['valid'] and not prior['late']:
                if prior['source_version']!=active:raise ValueError('superseded source version')
                active=prior['version'];adopted.add(active)
            current=by_tick.get(tick)
            if current and (current['source_version']!=active or not can_commit(by_version[active]['end_tick'],tick)):
                raise ValueError('observation/committed prefix mismatch')
        if tick<len(rows):
            row=rows[tick]
            if row['tick']!=tick or row['active_version']!=active:raise ValueError('actual epoch/calendar mismatch')
            if not by_version[active]['start_tick']<=tick<by_version[active]['end_tick']:raise ValueError('actual law expired')
    if any(c['adopted']!=(c['version'] in adopted) for c in candidates):raise ValueError('adoption record mismatch')
    return {'adopted_versions':sorted(adopted),'rows_checked':len(rows),'candidates_checked':len(candidates)}

def verify(packet_dir,out):
    manifest,refs,nominal=probe.packet(packet_dir);out=pathlib.Path(out);run=probe.read(out/'run.json');audit_schedule(run);model=mujoco.MjModel.from_xml_path(manifest['model']['scene']);spec=mujoco.mjtState.mjSTATE_INTEGRATION
    gids=[mujoco.mj_name2id(model,mujoco.mjtObj.mjOBJ_GEOM,n) for n in ('FR','FL','RR','RL')]
    nominal_at=lambda tick:dict(refs[tick%70],qpos=np.asarray(refs[tick%70]['qpos'])+np.r_[(tick//70)*nominal['command_vx']*nominal['period_s'],np.zeros(18)])
    candidates={c['version']:c for c in run['candidates']}
    def law(version,tick):
        if version==0:
            if not 0<=tick<6:raise ValueError('bootstrap extrapolation')
            return nominal_at(tick)
        c=candidates[version]
        if not c['start_tick']<=tick<c['end_tick']:raise ValueError('published law extrapolation')
        return next(s for s in c['stages'] if s['tick']==tick)
    d=mujoco.MjData(model);mujoco.mj_setState(model,d,np.asarray(run['initial_integration_state']),spec)
    states={0:np.asarray(run['initial_integration_state'])};max_state=max_force=max_law=0.;violations=[];observations=[];active=0;schedule_checks=[]
    with open('/tmp/go2_mujoco_experiment.lock','a') as lock:
        fcntl.flock(lock,fcntl.LOCK_EX|fcntl.LOCK_NB)
        for row in run['rows']:
            tick=row['tick']
            if tick and tick%6==0:
                previous=next((c for c in run['candidates'] if c['observation_tick']==tick-6),None)
                if previous and previous['valid'] and not previous['late'] and previous['source_version']==active: active=previous['version']
            if row['active_version']!=active:raise ValueError('publish ordering/version mismatch')
            requested=probe.control(model,law(active,tick),d);max_law=max(max_law,float(max(abs(requested-np.asarray(row['tau'])))))
            f=probe.step(model,d,np.asarray(row['tau']),gids);x=np.empty(len(run['initial_integration_state']));mujoco.mj_getState(model,d,x,spec)
            states[tick+1]=x.copy();max_state=max(max_state,float(max(abs(x-np.asarray(row['next_integration_state'])))));max_force=max(max_force,float(max(abs(f-np.asarray(row['forces'])))))
            if max(f)>180 or max(abs(np.asarray(row['tau'])))>35:violations.append(tick)
            observations.append({'tick':tick,'time':float(d.time),'qpos':d.qpos.tolist(),'qvel':d.qvel.tolist(),'forces':f.tolist()})
        for c in run['candidates']:
            source=c['source_version'];end=6 if source==0 else candidates[source]['end_tick'];obs=c['observation_tick']
            if not can_commit(end,obs):raise ValueError('candidate generated without complete committed prefix')
            if c['late']!=(c['generation_ms']>12) or (c['adopted'] and c['late']):raise ValueError('late candidate admitted')
            if obs not in states or not np.array_equal(states[obs],np.asarray(c['initial_integration_state'])):raise ValueError('candidate observation differs from continuous actual integration state')
            p=mujoco.MjData(model);mujoco.mj_setState(model,p,np.asarray(c['initial_integration_state']),spec);prefix_error=force_error=qv_error=0.
            stages=c['stages'] if c['valid'] or c['failure'].startswith('committed_prefix_force_violation:') else c['stages'][:-1]
            for j,s in enumerate(stages):
                qv_error=max(qv_error,float(max(abs(p.qpos-np.asarray(s['qpos'])))),float(max(abs(p.qvel-np.asarray(s['qvel'])))))
                if j<6:prefix_error=max(prefix_error,float(max(abs(probe.control(model,law(source,s['tick']),p)-np.asarray(s['tau'])))))
                f=probe.step(model,p,np.asarray(s['tau']),gids);force_error=max(force_error,float(max(abs(f-np.asarray(s['forces'])))))
            schedule_checks.append({'version':c['version'],'valid':c['valid'],'late':c['late'],'adopted':c['adopted'],'committed_prefix_law_error_nm':prefix_error,'candidate_force_replay_delta_n':force_error,'candidate_qv_replay_delta':qv_error})
    latency=np.asarray([c['generation_ms'] for c in run['candidates']]);report={'scope':'one initialized offline continuous run; synchronous compute with measured-late schedule, no liveRT or B1 certificate','target_steps':1400,'completed_steps':run['completed_steps'],'failure':run['failure'],'failure_tick':run['failure_tick'],'actual_integration_replay_delta':max_state,'actual_force_replay_delta_n':max_force,'actual_published_law_error_nm':max_law,'actual_constraint_violating_ticks':violations,'max_actual_force_n':max((max(row['forces']) for row in run['rows']),default=0),'candidate_count':len(candidates),'late_candidates':[c['version'] for c in run['candidates'] if c['late']],'adopted_candidates':sum(c['adopted'] for c in run['candidates']),'generation_ms':{'min':float(min(latency)),'median':float(np.median(latency)),'max':float(max(latency))},'candidate_checks':schedule_checks,'observations':observations,'run_sha256':probe.sha(out/'run.json')}
    probe.write(out/'verification.json',report);print({k:v for k,v in report.items() if k not in ('candidate_checks','observations')})
if __name__=='__main__':
    p=argparse.ArgumentParser();p.add_argument('mode',choices=('prepare','verify'));p.add_argument('packet_dir');p.add_argument('out_dir');a=p.parse_args();(prepare if a.mode=='prepare' else verify)(a.packet_dir,a.out_dir)
